diff --git a/op_tests/triton_tests/moe/test_moe_gemm_a16w4.py b/op_tests/triton_tests/moe/test_moe_gemm_a16w4.py
index 4ed667035c3..f789bb14745 100644
--- a/op_tests/triton_tests/moe/test_moe_gemm_a16w4.py
+++ b/op_tests/triton_tests/moe/test_moe_gemm_a16w4.py
@@ -282,17 +282,9 @@ def test_op(
     w_tri, w_scale_tri = downcast_to_mxfp(w_tri, weight_dtype, axis=1)
     w_ref = upcast_from_mxfp(w_tri, w_scale_tri, torch.bfloat16, axis=1)
     if hbm_swizzling:
-        if arch_info.get_arch() == "gfx1250":
-            swizzle_mx_scale = "GFX1250_SCALE"
-            w_scale_tri = shuffle_scale_moe(
-                w_scale_tri, arch="gfx1250", preshuffle_factor=32, scale_kwidth=8
-            )
-        else:
-            assert arch_info.get_arch() == "gfx950"
-            swizzle_mx_scale = "CDNA4_SCALE"
-            w_scale_tri = shuffle_scale_moe(
-                w_scale_tri, arch="gfx950", preshuffle_factor=32, scale_kwidth=8
-            )
+        w_scale_tri, swizzle_mx_scale = shuffle_scale_moe(
+            w_scale_tri, preshuffle_factor=32, scale_kwidth=8, return_layout=True
+        )
     else:
         swizzle_mx_scale = None
 
diff --git a/op_tests/triton_tests/moe/test_moe_gemm_a4w4.py b/op_tests/triton_tests/moe/test_moe_gemm_a4w4.py
index 37566aa6f4e..f646b310515 100644
--- a/op_tests/triton_tests/moe/test_moe_gemm_a4w4.py
+++ b/op_tests/triton_tests/moe/test_moe_gemm_a4w4.py
@@ -318,9 +318,12 @@ def test_op(
             swizzle_mx_scale = "GFX1250_SCALE"
             w_scale_tri = preshuffle_moe_wscale(w_scale_tri)
         elif get_arch() == "gfx950":
-            swizzle_mx_scale = "CDNA4_SCALE"
-            w_scale_tri = shuffle_scale_moe(
-                w_scale_tri, arch="gfx950", preshuffle_factor=32, scale_kwidth=8
+            w_scale_tri, swizzle_mx_scale = shuffle_scale_moe(
+                w_scale_tri,
+                arch="gfx950",
+                preshuffle_factor=32,
+                scale_kwidth=8,
+                return_layout=True,
             )
         else:
             assert False, "Unsupported architecture"
diff --git a/op_tests/triton_tests/moe/test_moe_gemm_a8w4.py b/op_tests/triton_tests/moe/test_moe_gemm_a8w4.py
index 2e2ef672099..c3fc8d5e07f 100644
--- a/op_tests/triton_tests/moe/test_moe_gemm_a8w4.py
+++ b/op_tests/triton_tests/moe/test_moe_gemm_a8w4.py
@@ -348,9 +348,12 @@ def test_op(
             w_scale_tri = preshuffle_moe_wscale(w_scale_tri)
         else:
             assert get_arch() == "gfx950"
-            swizzle_mx_scale = "CDNA4_SCALE"
-            w_scale_tri = shuffle_scale_moe(
-                w_scale_tri, arch="gfx950", preshuffle_factor=32, scale_kwidth=8
+            w_scale_tri, swizzle_mx_scale = shuffle_scale_moe(
+                w_scale_tri,
+                arch="gfx950",
+                preshuffle_factor=32,
+                scale_kwidth=8,
+                return_layout=True,
             )
     else:
         swizzle_mx_scale = None
diff --git a/op_tests/triton_tests/moe/test_moe_gemm_a8w8.py b/op_tests/triton_tests/moe/test_moe_gemm_a8w8.py
index f0dbdda6e8a..dfcc94e6c47 100644
--- a/op_tests/triton_tests/moe/test_moe_gemm_a8w8.py
+++ b/op_tests/triton_tests/moe/test_moe_gemm_a8w8.py
@@ -357,9 +357,12 @@ def test_op(
         w_ref = upcast_from_mxfp(w_tri, w_scale_tri, torch.bfloat16, axis=1)
         w_static_scale = None
         if hbm_swizzling:
-            swizzle_mx_scale = "CDNA4_SCALE"
-            w_scale_tri = shuffle_scale_moe(
-                w_scale_tri, arch="gfx950", preshuffle_factor=32, scale_kwidth=8
+            w_scale_tri, swizzle_mx_scale = shuffle_scale_moe(
+                w_scale_tri,
+                arch="gfx950",
+                preshuffle_factor=32,
+                scale_kwidth=8,
+                return_layout=True,
             )
         else:
             swizzle_mx_scale = None
